### ISSUE TITLE
RFC: unify output of moby tool / input of build images

### ISSUE THREAD

**Issue by deitch** [linuxkit/linuxkit#2469]
Right now, we have two distinct first-stage output formats from `moby build`, both of them tar files:

* separate: laid out like on a filesystem, primarily for use with `iso` or other immutable formats, with `cmdline` and `kernel` in `boot/`
* merged: 3 files in the root of the tar: `kernel`, `initrd.img`, `cmdline`

I don't know if others hit this, but I often end up getting confused when working on build formats, as the output of `tar` or the input of what is expected may vary.

Proposal: we _always_ have the output of `tar` (and the intermediate output to any modifiers like `mk-raw`, `mk-iso`, etc.) be one format. Either it always is 3 files and let the modifier stage open up and empty the `initrd.img` and move the `kernel` and `cmdline` into `boot/`; or it is always the full layout, and let the modifier stage build an `initrd.img` and move the `kernel` and `cmdline` out of `boot/` into the root.

In any case, I find the code in `tool/` somewhat confusing as it has to pick which output image.

Thoughts?

**Comment by justincormack** [comment]
This issues should probably be on `moby/tool`. This is somewhat inconsistent, the main reason is that we want to use the same initrd on all the versions that use one. We could move to having each image construct an initrd if it needs one; it used to be the case that everything needed it of course.

**Comment by deitch** [comment]
> This issues should probably be on moby/tool

I was thinking about that. It impacts both places: `moby/tool` creates the output, while the packages here consume them and create output.

> the main reason is that we want to use the same initrd on all the versions that use one.

I can see that. In that case, output `initrd.img`, and let formats that expect raw filesystems (mainly `iso`) extract them.

**Comment by justincormack** [comment]
That has downsides - in particular the translation to `cpio` is a bit lossy, eg hard links are replaced with symlinks due to the rather broken support in `cpio`. We could provide the initramfs etc inside the tarball and have the things that dont need it delete it.

**Comment by justincormack** [comment]
Related to this, I was thinking of adding a generic extension to allow you to specify your own docker/linuxkit images to run these conversions as well as the built in ones, to make iteration easier.

**Comment by deitch** [comment]
That would be nice. What I do is:

1. `moby build -output tar linuxkit.yml`
2. `cat linuxkit.yml | docker run -I --rm <my_image> > <myoutput>`

Of course, since `-output tar` sometimes gives one kind of tar layout and sometimes another, it gets messy (which is why I opened this issue....)

**Comment by justincormack** [comment]
`-output tar` always does the same thing, confused....

**Comment by deitch** [comment]
> -output tar always does the same thing

It does. But when you run with different `-output`, the streamed tar file (which is input to a post-processor) sometimes matches the output of `-output tar`, and sometimes not.

`-output tar`, I believe, gives an expanded filesystem. If you do several others, it runs through `tarToInitrd`, then streams that out.

Basically, I cannot do `-output tar` and expect to `cat linuxkit.tar` and have it match every post-processor. Some work with that, some do not.

**Comment by justincormack** [comment]
yes, there should probably be a suitable output for the other post processors ie `output tar-initrd`.

**Comment by deitch** [comment]
That might work too. I think it is cleaner to just always put out and initrd and have the iso or others expand it, but whatever works.
